src/main/java/com/example/weatherwaveapi/service/WeatherService.java

```diff
@@ -1,21 +1,87 @@
+import org.springframework.web.client.HttpClientErrorException;
+import org.springframework.web.client.HttpServerErrorException;
 import org.springframework.web.client.RestTemplate;
+import org.springframework.web.util.UriComponentsBuilder;
 
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.springframework.http.HttpMethod.GET;
+
+@RequiredArgsConstructor
 @Service
 @Slf4j
-@RequiredArgsConstructor
 public class WeatherService {
-    private final OpenWeatherUrlBuilder urlBuilder;
+    private static final String EXCEPTION_MESSAGE = "HTTP error occurred while processing request. Exception message: {}";
+    private static final String ERROR_MESSAGE = "Failed to retrieve weather data for city ";
+    private static final String STRING_MESSAGE_FORMAT = "Search string: %s and error: %s";
+
     private final RestTemplate restTemplate;
+    private final GeneralSettings generalSettings;
+
+    public WeatherOpenApiContainer getWeatherBySelectedCity(String city) {
+        String url = getUrlForSelectedCity(city);
+        try {
+            ResponseEntity<WeatherOpenApiContainer> responseEntity = restTemplate.exchange(
+                    url,
+                    GET,
+                    new HttpEntity<>(new HttpHeaders()),
+                    new ParameterizedTypeReference<>() {
+                    }
+            );
+            return responseEntity.getBody();
 
-    public Object getWeatherBySelectedCity(String city) {
-        String url = urlBuilder.buildWeatherUrl(city);
-        return restTemplate.getForObject(url, Object.class);
+        } catch (HttpClientErrorException | HttpServerErrorException e) {
+            log.error(EXCEPTION_MESSAGE, e.getMessage());
+            return WeatherOpenApiContainer.builder()
+                    .success(false)
+                    .errorMessage(String.format(STRING_MESSAGE_FORMAT, city, ERROR_MESSAGE))
+                    .build();
+        }
     }
-}
 
+    public WeatherResponse getWeather(WeatherRequest request) {
+
+        List<WeatherApiResponse> weatherResponses = new ArrayList<>();
+        for (String city : request.cities()) {
+            WeatherOpenApiContainer weatherApiResponse = getWeatherBySelectedCity(city);
+            weatherResponses.add(convertContainer(weatherApiResponse));
+        }
+
+        return WeatherResponse.builder()
+                .responses(weatherResponses)
+                .build();
+    }
+
+    private String getUrlForSelectedCity(String city) {
+        UriComponentsBuilder builder = UriComponentsBuilder.fromHttpUrl(generalSettings.getOpenWeatherApi().weatherUrl())
+                .queryParam("q", city)
+                .queryParam("appid", generalSettings.getOpenWeatherApi().key());
+
+        return builder.toUriString();
+    }
+
+    private WeatherApiResponse convertContainer(WeatherOpenApiContainer container) {
+        return WeatherApiResponse.builder()
+                .success(true)
```

Comment:
let's take this value from the WeatherOpenApiContainer

Reply:
I tried it but in that case it was null, so I just removed it